Adds compression status header to compressor_filter

pilifjed opened the pull request:
Commit Message: Adds compression status header to compressor_filter
Additional Description:

This change introduces a new response header x-envoy-compression-status to the compressor_filter. This header provides information on whether the response was compressed and, if not, the reason why compression was skipped.

The possible values for the status include:

Compressed: Indicates the response was compressed by this filter. Includes an additional OriginalLength parameter which stores the value from the ContentLength header that is removed on compression.
ContentLengthTooSmall: Skipped due to content length being below the minimum threshold.
ContentTypeNotAllowed: Skipped because the content type is not configured for compression.
EtagNotAllowed: Skipped due to the presence of an ETag header when disable_on_etag_header is true.
StatusCodeNotAllowed: Skipped because the response status code is not configured for compression.
The value format is <encoder-type>;<status>[;<additional-params>]. Multiple compressor filters in a chain will append their results separated by commas.

To enable this header, the runtime feature envoy.reloadable_features.compressor_add_status_header must be true. When enabled, the order of checks within the filter is slightly modified to emit the appropriate status reason. The behavior with the flag disabled remains unchanged.

Risk Level: Medium
Testing:

Added new unit tests in compressor_filter_test.cc and integration tests in compressor_filter_integration_test.cc to validate the presence and content of the x-envoy-compression-status header for various scenarios.
Adjusted existing tests to run with the runtime flag envoy.reloadable_features.compressor_add_status_header both enabled and disabled to ensure no behavioral changes for the existing logic.
Docs Changes: added section on the header to the: docs/root/configuration/http/http_filters/compressor_filter.rst
Release Notes: compressor: added the x-envoy-compression-status response header, which details if compression was applied or why it was skipped. This feature is gated by the envoy.reloadable_features.compressor_add_status_header runtime guard.
Platform Specific Features: N/A
Runtime guard: envoy.reloadable_features.compressor_add_status_header

mattklein123 reviewed (changes requested):
At a high level this seems fine to me but 2 things:
1) I would make this a proper configuration option and not a runtime switch.
2) Needs a release note.

/wait

pilifjed commented:
@mattklein123 Everything should be good now. Could you have a second look?

KBaichoo commented:
PTAL @mattklein123 

@pilifjed can you merge main?

mattklein123 reviewed (dismissed):
Thanks!

mattklein123 commented:
Sorry needs main merge.

/wait

pilifjed commented:
/retest

pilifjed commented:
@mattklein123 I rebased on top of main and squashed the commits locally. Minor adjustments were needed so listing those here for your convenience: 

- two new tests were added in the main commits so I adjusted them to test both configuration flag enabled and disabled:
  - PerRouteCompressorLibraryOverride 
  - PerRouteCompressorLibraryConfigCreation
- adjusted the `compressor filter.cc` to use `getContentEncoding()` method in the configuration flag enabled code branch to match the change in the configuration disabled flag. (PerRouteCompressorLibraryOverride test with flag enabled tests this behavior)

The rest of the changes were approved before rebase. Could you take a look at those and approve if those work for you?

pilifjed commented:
/retest
